File: Data Structures & Algorithms/insert-new-interval/submission-3.py
class Solution:
    def insert(self, intervals: List[List[int]], newInterval: List[int]) -> List[List[int]]:
        stack = []
        isNewIntervalInserted = False
        for i, interval in enumerate(intervals):
            #if overlaps, merge interval and new interval
            if interval[0] <=  newInterval[1] and interval[1] >= newInterval[0]:
                newInterval = [min(interval[0], newInterval[0]), max(interval[1], newInterval[1])]
            #if before new interval, append:
            elif interval[1] <= newInterval[0]:
                stack.append(interval)
            #if after new interval, append newInterval and rest of the intervals
            else: 
                stack.append(newInterval)
                isNewIntervalInserted = True
                stack += (intervals[i:])
                break
        if not isNewIntervalInserted:
            stack.append(newInterval)
        return stack

File: Data Structures & Algorithms/merge-intervals/submission-3.java
class Solution {
    public int[][] merge(int[][] intervals) {
        Arrays.sort(intervals, (a, b) -> {
            if(a[0] == b[0]) return Integer.compare(a[1], b[1]);
            return Integer.compare(a[0], b[0]);
        });

        ArrayDeque<int[]> stack = new ArrayDeque<>();
        for (int[] interval: intervals) {
            if (!stack.isEmpty() && stack.peek()[1] >= interval[0]) {
                int[] prevInterval = stack.peek();
                stack.pop();
                int[] newInterval = {prevInterval[0], Math.max(prevInterval[1], interval[1])};
                stack.push(newInterval);
            } else {
                stack.push(interval);
            }
        }

        return stack.toArray(new int[0][]);
    }
}
